src/decompiler.cc

```diff
@@ -0,0 +1,426 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/decompiler.h"
+
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <cinttypes>
+#include <cstdarg>
+#include <cstdio>
+#include <iterator>
+#include <map>
+#include <numeric>
+#include <string>
+#include <vector>
+#include <sstream>
+
+#include "src/cast.h"
+#include "src/common.h"
+#include "src/expr-visitor.h"
+#include "src/ir.h"
+#include "src/ir-util.h"
+#include "src/literal.h"
+#include "src/stream.h"
+
+#define WABT_TRACING 0
+#include "src/tracing.h"
+
+namespace wabt {
+
+namespace {
+
+struct Decompiler : ModuleContext {
+  Decompiler(Stream& stream, const Module& module,
+             const DecompileOptions& options)
+      : ModuleContext(module), stream(stream), options(options) {}
+
+  struct Value {
+    std::vector<std::string> v;
+    bool needs_bracketing;  // TODO: replace with a system based on precedence?
+
+    size_t width() {
+      size_t w = 0;
+      for (auto &s : v) {
+        w = std::max(w, s.size());
+      }
+      return w;
+    }
+  };
+
+  std::string Indent(size_t amount) {
+    return std::string(amount, ' ');
+  }
+
+  string_view OpcodeToToken(Opcode opcode) {
+    return opcode.GetDecomp();
+  }
+
+  void IndentValue(Value &val, size_t amount, string_view first_indent) {
+    WABT_TRACE_ARGS(IndentValue, "\"" PRIstringview "\" - %d",
+                    WABT_PRINTF_STRING_VIEW_ARG(val.v[0]), val.v.size());
+    auto indent = Indent(amount);
+    for (auto& s : val.v) {
+      auto is = &s != &val.v[0] || first_indent.empty()
+                ? string_view(indent)
+                : first_indent;
+      s.insert(0, is.data(), is.size());
+    }
+  }
+
+  void WrapChild(Value &child, string_view prefix, string_view postfix) {
+    WABT_TRACE_ARGS(WrapChild, "\"" PRIstringview "\" - \"" PRIstringview "\"",
+                    WABT_PRINTF_STRING_VIEW_ARG(prefix),
+                    WABT_PRINTF_STRING_VIEW_ARG(postfix));
+    auto width = prefix.size() + postfix.size() + child.width();
+    auto &v = child.v;
+    if (width < target_exp_width ||
+        (prefix.size() <= indent_amount && postfix.size() <= indent_amount)) {
+      if (v.size() == 1) {
+      // Fits in a single line.
+        v[0].insert(0, prefix.data(), prefix.size());
+        v[0].append(postfix.data(), postfix.size());
+      } else {
+        // Multiline, but with prefix on same line.
+        IndentValue(child, prefix.size(), prefix);
+        v.back().append(postfix.data(), postfix.size());
+      }
+    } else {
+      // Multiline with prefix on its own line.
+      IndentValue(child, indent_amount, {});
+      v.insert(v.begin(), std::string(prefix));
+      v.back().append(postfix.data(), postfix.size());
+    }
+  }
+
+  void BracketIfNeeded(Value &val) {
```

Comment:
So we decouple creating bracketed expressions from actually bracketing them, to avoid brackets when they aren't needed. Nice.
Possibly worth a comment in the struct definition?